### DISCUSSION TITLE
Signature formats

### DISCUSSION THREAD

**Discussion by lann** [bytecodealliance/SIG-Registries#17] [General]
We have so far sidestepped the question of how to encode signatures in registry APIs, but it will need to be addressed eventually. Let's start that discussion here.

**Comment by lann** [comment]
I'll enumerate what I see as the most viable options for encoding signatures:

* **JSON Web Signatures** ([RFC 7515](https://datatracker.ietf.org/doc/html/rfc7515))
  * :+1: Used in the popular JWT format; many existing implementations
  * :-1: [Infamous for implementation vulnerabilities](https://paragonie.com/blog/2017/03/jwt-json-web-tokens-is-bad-standard-that-everyone-should-avoid) (mostly related to its crypto agility design)

* **[DSSE](https://github.com/secure-systems-lab/dsse)**
  * :+1:/:-1: Simple design / small feature set
  * :+1: Increasing adoption in supply chain security software (in-toto, TUF, sigstore)

* **Custom**
  * :+1: Can focus on our relatively simple requirements
  * :-1: Must be implemented for every client language

**Comment by npmccallum** [comment]
As promised on the call, I'll provide some background on the claim that JWS is "infamous for implementation vulnerabilities."

# Opinions are Free

I'll start by noting that these are my opinions.  However, I've worked in cryptography for the last 12 years and I'm one of the inventors of the McCallum-Relyea key exchange. So my opinions don't come without the benefit of some experience.

The provided link criticizing JOSE contains a number of opinions. For example:

1. It claims that "Most cryptographers recommend migrating away from RSA." In actuality, the cryptographic community is deeply divided on this topic, [with many advocating for the use of RSA as having a better chance at surviving factorization under Shor's algorithm on a Quantum computer](https://security.stackexchange.com/questions/33069/why-is-ecc-more-vulnerable-than-rsa-in-a-post-quantum-world). This is, BTW, the opinion of NIST, who has recommended not to transition from RSA to elliptic curves.

2. It recommends that "Public-key cryptography should be avoided if possible." There is some merit to this claim in that we have no public-key cryptography solutions that are provably quantum resistant. However, this recommendation is not a very practical one for the majority of cryptographic uses. Following this advice would eliminate that vast majority of the cryptographic industry.

3. It recommends none of the NIST algorithms. And while there may be some legitimate concerns about the NIST algorithms, they are the only algorithms which are usable in any regulated industry. Further, the only recommended public-key algorithms by this post are all created by the same person (the [in]famous DJ Bernstein). I support the use of these algorithms, but they are not yet approved by NIST/FIPS and are unsuitable for a number of political contexts.

The author does have some legitimate concerns. However, the post broadly overstates these problems.

# JOSE & JWS

The [JOSE RFCs](https://datatracker.ietf.org/group/jose/documents/) define a general purpose system for cryptography data formats, specifically:

* JSON Web Keys (JWK)
* JSON Web Signatures (JWS)
* JSON Web Encryption (JWE)
* JSON Web Token (JWT)

While no specification is perfect, JOSE now has a large number of mature implementations. We are concerned with JWS specifically.

In particular, the JOSE crypto-system was seen as a great simplification of previous crypto-systems (most notably, X.509 certificates). This "lowering of the bar" encouraged many inexperience people to try their hand at writing cryptographic validators; a task notoriously fraught with difficulty.

# The Infamous JWS "Problem"

This is where this post gets a bit technical. Both JWK and JWS have an `alg` field. For JWK, this indicates which algorithm the key should be used with. For JWS, this indicates which algorithm was used to produce the signature. I'll show a concrete example of how these fields are used and why a problem occurred.

In fact, there wasn't one problem. There was a variety of problems that tripped up various implementations (but not all). Some of these problems weren't actually at the crypto-system level at all, but only occurred in systems implementing JWTs.

## A Concrete Example

Let's say we have a JWK that looks like this (notice the `alg` field):

```json
{
    "kty":"EC",
    "crv":"P-256",
    "alg":"ES256",
    "x":"f83OJ3D2xF1Bg8vub9tLe1gHMzV76e8Tus9uPHvRVEU",
    "y":"x_FEzRu9m36HLN_tue659LNpXW6pCyStikYjKIWI5a0",
    "d":"jpsQnnGQmL-YBIffH1136cspYG6-0iY7X1fCE9-E9LI"
}
```

Now, we want to sign the payload `foo` and produce a JWS. The result SHOULD look like this:

```json
{
    "payload":"Zm9v",
    "protected":"eyJhbGciOiJFUzI1NiJ9",
    "signature":"IVRgI8LnfBjcr2uWFKAvP0_lfmLHYkn9LBdK6cjPiq7yIMn2yVg60-LgKfWwPDiz06_50VsYIGK-UfkR_VCLfw"
}
```

(The above is the actual output of the C jose implementation.)

1. The `payload` field is the base64url encoding of `foo`.
2. The `protected` field is the base64url encoding of the following metadata: `{"alg":"ES256"}`.

Importantly, the `protected` field is included in the signature. This means that if you attempt to change the `alg` field, the signature will fail to validate. Or should...

## The Problems

### The "none" Algorithm

The first problem arises in that the RFC defines a `none` algorithm. They probably shouldn't have. However, this used to be common in the cryptography space to allow testing of the container format. It has (rightfully) fallen into disrepute over time because unsuspecting engineers can make mistakes with it. JOSE was standardized right as this was happening, so it is mostly a case of unfortunate timing.

Therefore, some JOSE implementations wrote their signature validator to always succeed if the signature algorithm was `none`. This is a bit like checking to see if a door is unlocked and reporting "no" because the door is wide open.

It is regrettable that JOSE has a `none` algorithm. But on the other hand, allowing this behavior could only happen because inexperienced people were writing crypto code (don't write your own crypto code). JWS now has many mature implementations for this in most languages that are not vulnerable to this problem.

Note that RFC 7515 precludes this behavior when it states (4.1.1): 

```
The JWS Signature value is not valid if the "alg" value does not represent
a supported algorithm or if there is not a key for use with that algorithm
associated with the party that digitally signed or MACed the content.
```

### The Protected/Unprotected Header Problem

The JWS specification wanted the ability to attach mutable metadata to a signature. You could express metadata as either `protected` or as "unprotected" in the `header` field. The `protected` headers were immutable since they were included in the signature and the "unprotected" headers were mutable.

The problem that arose here is that naive implementors would treat the `alg` field as valid when it was an "unprotected" header. This allowed attackers to change the `alg` field and trick the validator into processing other algorithms.

This precise attack is discussed at length in the `Security Considerations - Algorithm Protection` section (10.7).

### The Public/Private Problem

Some implementations likewise made another naive mistake. If the attacker submitted a signature with `"alg": "HS256"` they would treat the bytes of the public key of the elliptic curve as the private, symmetric HMAC key. This meant that since the attacker always knew the public key material she could always forge a signature.

This mistake might be encouraged in a minor way by the presence of the `alg` field in the signature, but it is greatly compounded by a variable reuse problem in the implementation. No specification should have to outline that you should never treat a public key like a private key. Nor should you reuse variables between public and private keys.

### The Canonical JSON Problem

Some others have made the claim that the lack of a canonical encoding for JSON is a problem for JOSE. I'm aware this is a theoretical problem. However, I'm unaware of any vulnerability which has been caused by this.

### Summary

One or more of these first three problems hit a number of JOSE implementations. Because they are all tenuously related to the `alg` header, some people conflated these three issues into one larger issue and proclaimed the JOSE specifications as failed.

Yet, in retrospect, the real problem is that JOSE made cryptography approachable for the first time. The shift from ASN.1 encoding to JSON encoding meant that people who had no experience with cryptography were encouraged to try their hand at implementing it. Therefore, we saw dozens of implementations arise very quickly with varying degrees of quality. Few were rigorously tested or reviewed.

While the wisdom of putting `alg` as metadata can be debated (it is certainly useful in development, debugging and auditing scenarios; with some corresponding risks for implementors), the far bigger problem is that security-critical systems were built using basically unaudited crypto-systems written by people who were cryptography novices.

# The "Solution"

While the JOSE specification isn't perfect, the "fixes" are, IMHO, worse.

## Maturity

Both DSSE and PASETO claim to avoid the implementation pitfalls "caused" by JOSE. Yet, because they aren't broadly adopted and mature, more people will be forced to implement them from scratch. This will lead to immature implementations with new security issues. Cryptography is hard.

OTOH, if we use JOSE, users are most likely going to pick an off-the-shelf implementation for their own language which has already been tried in the real world with security fixes applied.

## Simplicity

Both DSSE and PASETO claim to be simpler. But the "simplicity delta" between JOSE and DSSE/PASETO is orders of magnitude smaller than the delta between X.509/GNUPG and JOSE. While both will gain some limited usage, JOSE is "simple enough" for most use cases and also provides sufficient flexibility as our needs grow.

## The Custom Option

I really don't see a compelling case for developing our own format. I won't deal with it here. However, I will note that we are focused on building something that is a "step up" from the current registry experience. I think it would benefit us to limit the scope of our innovations to things that truly provide value.

# Conclusion

All cryptography systems have a history of vulnerable implementations (the list of CVEs for OpenSSL is... long). While not perfect, JOSE enabled the rapid development of dozens of implementations in many languages in just a few short years (JOSE is only 7 years old; X.509 is decades old and has far fewer implementations). This burst of innovation caused some vulnerabilities with similar themes. But mature, secure implementations are now available in almost any language. In contrast, the competing "improvements" on JOSE offer minimal gains at the steep cost of reimplementation.

**Reply by lann** [reply]
Thanks for the writeup. I agree with much of this, but I'd like to respond to a couple of points:

> if we use JOSE, users are most likely going to pick an off-the-shelf implementation for their own language which has already been tried in the real world with security fixes applied.

This is true in many situations, but the availability of high quality implementations shouldn't be overstated, especially those exposing JWS separately from JWT. Anecdotally: a while ago I wrote an ACME (RFC 8555, the protocol used by Let's Encrypt) implementation, which [signs requests as JWS payloads](https://datatracker.ietf.org/doc/html/rfc8555#section-6.4). I couldn't find a Rust library that implemented all the required features without FFIing to OpenSSL (I was targeting Wasm), so I [wrote my own](https://github.com/localcert/acme-rs/blob/main/src/crypto/jws.rs). I _think_ it is correct, but I am not the last person who will re-implement it. Given the history of incorrect implementations, I think this is a real concern.

> JOSE is "simple enough" for most use cases

Subjectively, I disagree on this point. While a specific use of JOSE can be simple, fully-featured implementations are not; looking just at JWS there are, for example:
* multiple serialization forms (compact, flattened, "full")
* multiple key specification forms (JWK, JWKS, X.509)
* multiple varieties of header (protected vs unprotected, "critical")
* the `alg` header as discussed above

I place a high cost premium on complexity, and JOSE has more complexity than I am entirely comfortable with.

**Reply by npmccallum** [reply]
@lann 

1. I agree that Rust is an unfortunate lacuna in the state of JOSE implementations. However, I've already [started an implementation](https://github.com/profianinc/drawbridge/tree/main/crates/jose). It currently just contains serialization. But I'd be willing to work on this.

2. Regarding the complexity points, I don't think I was ever suggesting that we require full JOSE support. My intent was that we choose a relevant subset of JWS to support. This should give us something with a level of simplicity similar to DSSE, while giving us the benefits of fully specified behavior and off-the-shelf implementations.

**Comment by lukewagner** [comment]
Prodding @jedisct1 a bit, I was not able to elicit a strong opinion regarding the JOSE suite; Frank independently described most of the issues that were discussed above, culminating with the choice of JWS for signing being "reasonable".  In contrast, he wasn't already familiar with DSSE, which I suppose speaks to this general concern that it's not nearly as well-studied or widely-deployed as the JOSE standards.

Just to check assumptions: I see that the JOSE suite is described by IETF RFCs but I wasn't able to find any standard-ish body covering DSSE which, as a general rule, I count as a point in the JOSE suite's favor.  I was also curious if there were any Web standards that depended on JOSE (acting as a form of "anchor" that will keep JOSE alive and kicking for a "real long time"); all we were able to find was [W3C Verified Credentials](https://www.w3.org/TR/vc-data-model/) which, from a bit of poking around, seems to have some industry use and forward motion but isn't like, e.g., a WebSockets-level of permanence.

Independently, I was interested to know if Fastly already or in the future made native use of any of any of these standards.  We don't at this moment, but there are some features on the tentative roadmap that involve native support for JOSE, so that's another small data point.

These are all more "structural" points regarding this question (as opposed the technical question of whether DSSE satisfies all our current or future use cases or JOSE is over-complicated), but I think the structural points generally lean in favor of JOSE.

**Reply by lann** [reply]
I have no concerns with JOSE's adoption or ongoing maintenance; it is used in OpenID Connect. :slightly_smiling_face:

My hesitation here is that JOSE is more complex than warranted for our requirements. Many systems that use signatures don't adopt _any_ existing standard for signature encoding because it simply isn't necessary.

I recently came across this more in-depth review of signature formats including evaluations of JWS, DSSE, and "Nothing At All!": https://dlorenc.medium.com/signature-formats-9b7b2a127473

**Comment by lann** [comment]
Per meeting non-disagreement, I think we'll tentatively go with this: @npmccallum will contribute an implementation of a subset of JOSE to BCA (presumably subsets of JWS and perhaps JWK), and we will specify a small subset of those specs to be supported by the registry protocol.

**Reply by kulakowski-wasm** [reply]
Wonderful!

@npmccallum if you end up needing any help with administrative stuff (where to put code in the BCA github, in particular), I'm happy to help (mostly by escalating ;p )
